ql/indexes/equityindex.cpp

```diff
@@ -0,0 +1,100 @@
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+#include <ql/indexes/equityindex.hpp>
+#include <ql/settings.hpp>
+#include <utility>
+
+namespace QuantLib {
+
+    namespace {
+        Real resolveSpot(const Handle<Quote>& spot, Real todaysFixing) {
+            QL_REQUIRE(!spot.empty() || todaysFixing != Null<Real>(),
+                       "Cannot forecast equity index, missing both spot and historical index");
+            return spot.empty() ? todaysFixing : spot->value();
+        }
+    }
+
+    EquityIndex::EquityIndex(std::string name,
+                             Calendar fixingCalendar,
+                             Handle<YieldTermStructure> interest,
+                             Handle<YieldTermStructure> dividend,
+                             Handle<Quote> spot)
+    : name_(std::move(name)), settlementCalendar_(std::move(fixingCalendar)),
+      interest_(std::move(interest)), dividend_(std::move(dividend)), spot_(std::move(spot)) {
+
+        registerWith(interest_);
+        registerWith(dividend_);
+        registerWith(spot_);
+        registerWith(Settings::instance().evaluationDate());
+        registerWith(IndexManager::instance().notifier(EquityIndex::name()));
+    }
+
+    Real EquityIndex::fixing(const Date& fixingDate, bool forecastTodaysFixing) const {
+
+        QL_REQUIRE(isValidFixingDate(fixingDate), "Fixing date " << fixingDate << " is not valid");
+
+        Date today = Settings::instance().evaluationDate();
+
+        if (fixingDate > today || (fixingDate == today && forecastTodaysFixing))
+            return forecastFixing(fixingDate);
+
+        Real result = pastFixing(fixingDate);
+
+        if (result != Null<Real>())
+            // if historical fixing is present use it
+            return result;
+        
+        if (fixingDate == today && !spot_.empty())
+            // Today's fixing is missing, but spot is
+            // provided, so use it as proxy
+            return spot_->value();
+        
+        QL_FAIL("Missing " << name() << " fixing for " << fixingDate);
+    }
+
+    Real EquityIndex::pastFixing(const Date& fixingDate) const {
+        QL_REQUIRE(isValidFixingDate(fixingDate), fixingDate << " is not a valid fixing date");
+        return timeSeries()[fixingDate];
+    }
+
+    Real EquityIndex::forecastFixing(const Date& fixingDate) const {
+        QL_REQUIRE(!interest_.empty(),
+                   "null interest rate term structure set to this instance of " << name());
+
+        Date today = Settings::instance().evaluationDate();
+        today = settlementCalendar_.adjust(today, BusinessDayConvention::Preceding);
```

Comment:
Instead of modifying `today`, this could be a new variable `lastFixingDate` for clarity.